Prepare for the four channels.

frankban opened the pull request:
Improve how the channel argument is handled by subcommands.
Also deprecate the "development" channel, automatically translate it to "edge"
and guide the user to start using "edge".

frankban commented:
This can be QAed by compiling the 4channel branch of the charm store.
After that, run `make server` on the charm store, compile this branch in another
terminal tab in which you've exported JUJU_CHARMSTORE=http://localhost:8080
At this point you can start using the "charm" command and exercise the support for all
four channels: stable, candidate, beta and edge. You can also double check the
deprecation message when using the "development" channel.

jujugui commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
http://ci.jujugui.org:8080//job/charmstore-client/120/
Test PASSed.

rogpeppe commented on cmd/charm/charmcmd/cmd.go:
```diff
@@ -138,24 +139,56 @@ func newCharmStoreClient(ctxt *cmd.Context, auth authInfo) (*csClient, error) {
+
+// chanValue is a gnuflag.Value that stores a channel name
+// ("stable", "candidate", "beta", "edge" or "unpublished").
+type chanValue string
```
I think I'd be tempted to define this as:

```
type chanValue struct {
      C params.Channel
}
```

Then you won't need the Channel method,
and Set becomes a little simpler.

frankban replied:
Done.

rogpeppe commented on cmd/charm/charmcmd/attach.go:
```diff
@@ -11,18 +11,17 @@ import (
-	name    string
-	file    string
-	auth    authInfo
+	cv   chanValue
```
I think I'd still prefer "channel" as a field name here over cv, even though it's not strictly a params.Channel.

frankban replied:
Done.

rogpeppe commented on cmd/charm/charmcmd/cmd.go:
```diff
@@ -138,24 +139,56 @@ func newCharmStoreClient(ctxt *cmd.Context, auth authInfo) (*csClient, error) {
+// addChannelFlag adds the -c (--channel) flags to the given flag set.
+func addChannelFlag(f *gnuflag.FlagSet, cv *chanValue) {
+	chans := []string{
+		string(params.StableChannel),
```
I wonder if the params package should export a variable with all the channels in.
Perhaps it should export OrderedChannels and ValidChannels.

frankban replied:
Good idea! Done.

rogpeppe commented:
LGTM modulo a few suggestions, thanks!

jujugui commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
http://ci.jujugui.org:8080//job/charmstore-client/121/
Test PASSed.

frankban committed:
Prepare for the four channels.

Improve how the channel argument is handled by subcommands.
Also deprecate the "development" channel, automatically translate it to "edge"
and guide the user to start using "edge".

jujugui commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
http://ci.jujugui.org:8080//job/charmstore-client/122/
Test PASSed.

rogpeppe commented on cmd/charm/charmcmd/cmd.go:
```diff
@@ -138,24 +139,52 @@ func newCharmStoreClient(ctxt *cmd.Context, auth authInfo) (*csClient, error) {
 	return &csClient, nil
 }
 
+// addChannelFlag adds the -c (--channel) flags to the given flag set.
```
Perhaps document that channels is the set of allowed channels?

frankban replied:
Done.

rogpeppe commented on cmd/charm/charmcmd/publish.go:
```diff
@@ -95,19 +97,23 @@ var publishCharm = func(client *csclient.Client, id *charm.URL, channels []param
 
-	err = publishCharm(client.Client, c.id, []params.Channel{params.Channel(c.channel)}, c.resources)
+	channel := c.channel.C
+	if channel == params.NoChannel {
```
Rather than doing this, we could instantiate the chanValue to have StableChannel as its value. I think you'll then see it as the default when using --help (worth checking that I'm correct there thought)

frankban replied:
Ah! Didn't know that, and it works. Good suggestion.

rogpeppe commented:
LGTM with a couple of further minor suggestions, thanks!

jujugui commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
http://ci.jujugui.org:8080//job/charmstore-client/123/
Test PASSed.

frankban commented:
Thank you Roger for the review!
:shipit:

jujugui commented:
Status: merge request accepted. Url: http://ci.jujugui.org:8080/job/charmstore-client-merge
